### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by verenichyi**
update dev env and dev deps

### CODE REVIEW

File: src/tests/app.test.tsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    const defaultName = container.querySelector(`.birdName`);
+
+    expect(defaultImage.getAttribute('src')).toBe(`${image}`);
+    expect(defaultName).toHaveTextContent(/(\*)+/gi);
+
+    clickOnAllAnswers(container);
+
+    expect(defaultImage.getAttribute('src')).not.toBe(`${image}`);
+    expect(defaultName).not.toHaveTextContent(/(\*)+/gi);
+  });
```
Comment (julia-dream) [inline src/tests/app.test.tsx]: Давай разделим эти тесты каждый на несколько небольших в соответствии с общей логикой 
Для этого обычно используется describe в jest'e
здесь есть неплохое объяснение в сравнении с твоим вариантом https://stackoverflow.com/questions/32055287/what-is-the-difference-between-describe-and-it-in-jest

Comment (julia-dream) [reply src/tests/app.test.tsx]: Было бы замечательно, чтобы каждый it содержал в себе один, максимум два выражения expect

File: src/tests/app.test.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  test('Answer should render with different indicator after click', () => {
+    const { container } = render(<App />);
+    const indicator = container.querySelector(
+      'ul > li:nth-child(1) > button > div'
+    );
```
Comment (julia-dream) [inline src/tests/app.test.tsx]: Подобные константы было бы неплохо вынести в отдельный объект, и, возможно, шарить между тестами

File: src/tests/app.test.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    expect(indicator).toBeInTheDocument();
+    expect(indicator).toHaveClass('indicator default');
+
+    fireEvent.click(indicator);
+
+    expect(indicator).not.toHaveClass('indicator default');
```
Comment (julia-dream) [inline src/tests/app.test.tsx]: Вот здесь точно необходимо разделение на два теста - .toBeInTheDocument() и toHaveClass('indicator default');

и not.toHaveClass('indicator default'); после клика

Comment (julia-dream) [reply src/tests/app.test.tsx]: Тесты ведь нужны нам для того, чтобы поддерживать код и вносить изменения было гораздо проще - функционал проверят именно они 
А если у тебя сломается что-то при клике, то через конкретный тест это и можно будет увидеть
В текущей ситуации - если сломается присвоение класса индикатору - тест зафейлится - тебе, как разработчику, придется проверять каждый expect  в этом тесте, чтобы понять, что конкретно ломается

File: src/tests/app.test.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  test('Description component should has suggestion text before answer clicked', () => {
+    const { getByText } = render(<App />);
+
+    expect(getByText(/послушайте плеер/i)).toBeInTheDocument();
+    expect(getByText(/выберите птицу из списка/i)).toBeInTheDocument();
```
Comment (julia-dream) [inline src/tests/app.test.tsx]: Тоже в отдельные константы

File: src/tests/app.test.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    fireEvent.click(button);
+
+    expect(
+      container.querySelector(`section[class=description] img[alt=bird]`)
+    ).toBeInTheDocument();
+  });
+
+  test('Next level button should be disabled before right answer clicked and enabled after', () => {
+    const { container, getByText } = render(<App />);
```
Comment (julia-dream) [inline src/tests/app.test.tsx]: В каждом тесте ты используешь render(<App />);
Такое потворение логики должно было насторожить и вызвать желание вынести это на уровень выше 
Обычно используется  
```
beforeEach(() => {
   
  });
```

File: src/tests/app.test.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    const button = getByText(/next level/i);
+
+    expect(button).toBeInTheDocument();
+    expect(button).toBeDisabled();
+
+    clickOnAllAnswers(container);
+    expect(button).not.toBeDisabled();
```
Comment (julia-dream) [inline src/tests/app.test.tsx]: Аналогичную ситуацию с индикаторами описывала выше

File: src/tests/header.test.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import Header from 'src/view/components/Question/Header';
+import { nav } from 'src/constants/common';
```
Comment (julia-dream) [inline src/tests/header.test.tsx]: По поводу общей файловой структуры - существует и такой подход, как использовал ты - конкретная папка с тестами в отдельном месте
Но, в текущей ситуации, когда приложение четко разделено на компоненты, для которых ты и используешь тесты, стоит поместить файл с тестом непосредственно в папку с компонентом 
И, соответственно, ты будешь видеть, что для каждого компонента должны быть написаны тесты в соответствии с его функционалом

File: src/tests/header.test.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+describe('Header: ', () => {
+  let header;
+  let initialData;
+
+  beforeEach(() => {
+    initialData = {
+      score: 0,
+      level: 0,
+      nav,
+    };
```
Comment (julia-dream) [inline src/tests/header.test.tsx]: ты же не меняешь нигде initialData, эта константа может быть вынесена в принципе над блоком describe

File: src/tests/helpers.test.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import { formatTime, calculateTime } from 'src/utils/helpers';
+
+describe('Helper functions testing: ', () => {
+  test('Format time func should return doubled-value string', () => {
```
Comment (julia-dream) [inline src/tests/helpers.test.ts]: Выше оставляла ссылку, внутри блока describe лучше использовать it

Comment (julia-dream) [reply src/tests/helpers.test.ts]: Классическая структура теста для такой функции: 
```
describe (Format time func)
    describe (positive scenarios)
      it 
      it
    describe (negative scenarios)
      it 
      it
```
Каждый как положительный, так и негативный сценарий должен покрывать edge cases функции, и также проверять основной функционал

File: src/tests/helpers.test.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+  test(`Calculate time func should take time in seconds and return time in 'minutes : seconds' format`, () => {
+    expect(calculateTime(1000)).toBe(`16 : 40`);
+    expect(calculateTime(100)).toBe(`01 : 40`);
+    expect(calculateTime(100)).not.toBe(`1 : 40`);
+    expect(calculateTime(10)).toBe(`00 : 10`);
+    expect(calculateTime(0)).toBe(`00 : 00`);
+    expect(calculateTime(0)).not.toBe(`0 : 0`);
+  });
+});
```
Comment (julia-dream) [inline src/tests/helpers.test.ts]: Этот файл точно нужно переписать в соответствии с комментариями выше

**Review by julia-dream — CHANGES_REQUESTED** [review]
Класс!
Оставила несколько комментариев...

File: src/utils/helpers.test.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import { formatTime, calculateTime } from 'src/utils/helpers';
+
+describe('Helper functions testing: ', () => {
+  describe('calculateTime func: ', () => {
+    it(` should return time in 'doubled-value minutes : doubled-value seconds' format`, () => {
+      expect(calculateTime(10)).toBe(`00 : 10`);
+      expect(calculateTime(0)).toBe(`00 : 00`);
+    });
```
Comment (julia-dream) [inline src/utils/helpers.test.ts:8]: На будущее - было бы круто для таких функций проверять самый банальный вариант 'Throw error if paarameter type isn't a number' или что-то в таком духе

File: src/view/components/App/app.test.tsx [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    let name;
+
+    beforeEach(() => {
+      image = app.getByAltText(/bird/);
+      name = app.container.querySelector(testSelectors.questionBirdName);
+    });
+
+    it('has image and name', () => {
+      expect(image).toBeInTheDocument();
+      expect(name).toBeInTheDocument();
```
Comment (julia-dream) [inline src/view/components/App/app.test.tsx]: Можем, проверим на наличие в нужном месте, а не просто в документе?

File: src/view/components/App/app.test.tsx [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    });
+
+    describe('before a player chose the right answer', () => {
+      it('has default image', () => {
+        expect(image.getAttribute('src')).toBe(defaultImage);
+      });
+
+      it('has hidden name ', () => {
+        expect(name).toHaveTextContent(/(\*)+/gi);
```
Comment (julia-dream) [inline src/view/components/App/app.test.tsx]: регулярочку в константу

File: src/view/components/App/app.test.tsx [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        fireEvent.click(rightAnswerButton);
+      });
+
+      it('has certain image', () => {
+        expect(image.getAttribute('src')).not.toBe(defaultImage);
+      });
+
+      it('has certain name', () => {
+        expect(name).not.toHaveTextContent(/(\*)+/gi);
+      });
```
Comment (julia-dream) [inline src/view/components/App/app.test.tsx]: если есть notToBe, notToHave, то должна быть и положительная ситуация :)

File: src/view/components/App/app.test.tsx [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    it('should be disabled before right answer was clicked', () => {
+      expect(button).toBeDisabled();
+    });
+
+    it('should be enabled after right answer was clicked', () => {
+      fireEvent.click(rightAnswerButton);
+      expect(button).toBeEnabled();
+    });
+  });
+});
```
Comment (julia-dream) [inline src/view/components/App/app.test.tsx:32]: Мне казалось, мы обсуждали, что тесты стоить разделить отдельно для каждого компонента
А сейчас в одном файле и Description, и Answers и тд

**Review by julia-dream — CHANGES_REQUESTED** [review]

File: src/view/components/App/app.test.tsx (hunk 14 above)
Comment (verenichyi) [reply src/view/components/App/app.test.tsx]: так есть же :)

**Review by julia-dream — APPROVED** [review]
